Test/Optimizer/test_sr_logic.py

```diff
@@ -308,36 +312,110 @@ def mvt(v, f, params, samples, centered, w):
     assert tree_allclose(actual, expected)
 
 
-# TODO separate test for prepare_doks
+# TODO separate test for prepare_centered_oks
 @pytest.mark.parametrize("holomorphic", [True])
 @pytest.mark.parametrize("n_samp", [25, 1024])
 @pytest.mark.parametrize("jit", [True, False])
 @pytest.mark.parametrize(
     "outdtype, pardtype", r_r_test_types + c_c_test_types + r_c_test_types
 )
-def test_matvec_treemv(e, jit, holomorphic):
+def test_matvec_treemv(e, jit, holomorphic, pardtype, outdtype):
     diag_shift = 0.01
-    mv = sr_treemv_logic.mat_vec
-    pdoks = sr_treemv_logic.prepare_doks
+    mv = qgt_jacobian_pytree_logic._mat_vec
+
+    if not nkjax.is_complex_dtype(pardtype) and nkjax.is_complex_dtype(outdtype):
+        centered_jacobian_fun = qgt_jacobian_pytree_logic.centered_jacobian_cplx
+    else:
+        centered_jacobian_fun = qgt_jacobian_pytree_logic.centered_jacobian_real_holo
+
     if jit:
         mv = jax.jit(mv)
-        pdoks = jax.jit(pdoks, static_argnums=0)
+        centered_jacobian_fun = jax.jit(centered_jacobian_fun, static_argnums=0)
+
+    centered_oks = centered_jacobian_fun(e.f, e.params, e.samples)
+    centered_oks = qgt_jacobian_pytree_logic._divide_by_sqrt_n_samp(
+        centered_oks, e.samples
+    )
+    actual = mv(e.v, centered_oks)
+    expected = reassemble_complex(e.S_real @ e.v_real_flat, target=e.target)
+    assert tree_allclose(actual, expected)
+
 
-    doks = pdoks(e.f, e.params, e.samples)
-    actual = mv(e.v, doks, diag_shift)
+# TODO separate test for prepare_centered_oks
+# TODO test C->R ?
+@pytest.mark.parametrize("holomorphic", [True, False])
+@pytest.mark.parametrize("n_samp", [25, 1024])
+@pytest.mark.parametrize("jit", [True, False])
+@pytest.mark.parametrize("outdtype, pardtype", test_types)
+def test_matvec_treemv_modes(e, jit, holomorphic, pardtype, outdtype):
+    diag_shift = 0.01
+    model_state = {}
+    rescale_shift = False
+
+    def apply_fun(params, samples):
+        return e.f(params["params"], samples)
+
+    mv = qgt_jacobian_pytree_logic.mat_vec
+
+    homogeneous = pardtype is not None
+
+    if not nkjax.is_complex_dtype(outdtype):
+        mode = "real"
+    elif homogeneous and nkjax.is_complex_dtype(pardtype) and holomorphic:
+        mode = "holomorphic"
+    else:
+        mode = "complex"
+
+    if mode == "holomorphic":
+        v = e.v
+        reassemble = lambda x: x
+    else:
+        v, reassemble = nkjax.tree_to_real(e.v)
+
+    if jit:
+        mv = jax.jit(mv)
+
+    centered_oks, _ = qgt_jacobian_pytree_logic.prepare_centered_oks(
+        apply_fun, e.params, e.samples, model_state, mode, rescale_shift
+    )
+    actual = reassemble(mv(v, centered_oks, diag_shift))
     expected = reassemble_complex(
         e.S_real @ e.v_real_flat + diag_shift * e.v_real_flat, target=e.target
     )
     assert tree_allclose(actual, expected)
 
 
+@pytest.mark.parametrize("holomorphic", [True])
```

Comment:
that's not much of a parametrisation

Reply:
It's only supposed to test RR, RC and holomorphic CC, so it's enough to use the holomorphic example function